test-cases/longevity/longevity-lwt-24h-multidc.yaml

```diff
@@ -1,7 +1,7 @@
 test_duration: 1560
-prepare_write_cmd: ["cassandra-stress user profile=/tmp/c-s_lwt_big_data_multidc.yaml n=10000000 ops'(insert_query=1)' cl=QUORUM -mode native cql3 -rate threads=1000" ]
-stress_cmd: ["cassandra-stress user profile=/tmp/c-s_lwt_big_data_multidc.yaml ops'(lwt_update_one_column=1,lwt_update_two_columns=1,lwt_deletes=1)' cl=QUORUM duration=1350m -mode native cql3 -rate threads=800" ]
-stress_read_cmd: ["cassandra-stress user profile=/tmp/c-s_lwt_big_data_multidc.yaml ops'(select=1)' cl=SERIAL duration=1350m -mode native cql3 -rate threads=800" ]
+prepare_write_cmd: ["cassandra-stress user profile=/tmp/c-s_lwt_big_data_multidc.yaml n=10000000 ops'(insert_query=1)' cl=LOCAL_QUORUM serial-cl=LOCAL_SERIAL -mode native cql3 -rate threads=1000" ]
+stress_cmd: ["cassandra-stress user profile=/tmp/c-s_lwt_big_data_multidc.yaml ops'(lwt_update_one_column=1,lwt_update_two_columns=1,lwt_deletes=1)' cl=LOCAL_QUORUM serial-cl=LOCAL_SERIAL duration=1350m -mode native cql3 -rate threads=800" ]
+stress_read_cmd: ["cassandra-stress user profile=/tmp/c-s_lwt_big_data_multidc.yaml ops'(select=1)' cl=LOCAL_QUORUM serial-cl=LOCAL_SERIAL duration=1350m -mode native cql3 -rate threads=800" ]
 
 n_db_nodes: '4 3 2'
```

Comment:
In one of the 3 regions our `quorum` gets translated into `local all` meaning we cannot use any of those nodes for disruptive nemesis.
Are you sure about this change here?

Reply:
I am not sure I understand what do you mean. 
Scylla does not support "LOCAL_ALL" cl.
Why do you think we can not to use it for nemeses?

Reply:
quorum is quorum, and in the case of two 2 nodes, it would be one node.
I do agree it's not the best practice to have a region with only one node, and I'm not sure I remember the actual reason we are doing it like that in this case (@roydahan do you remember why ?)

Reply:
QUORUM of 2 is 2.
It's for the diversity of testing.

Reply:
> I am not sure I understand what do you mean. Scylla does not support "LOCAL_ALL" cl. Why do you think we can not to use it for nemeses?

I meant the idea 2/2 in one of the 3 regions. So, if we use nemesis against 1 of 2 nodes in that region our queries will fail to get 2 responses.

Reply:
That's a valid point against local_quorum.
@juliayakovlev how does it pass your testing?

Reply:
There is no problem. It passed with 2023.1 and 5.2

Reply:
let's scale the 3rd region to 3 nodes, node we can do nemesis safely on it

Reply:
Before doing it, I would like to understand if we had disruptive nemesis on a node that was on that region and if so, how didn't it fail the stress on that region.

Reply:
> Before doing it, I would like to understand if we had disruptive nemesis on a node that was on that region and if so, how didn't it fail the stress on that region.

WIll re-check

Reply:
> Before doing it, I would like to understand if we had disruptive nemesis on a node that was on that region and if so, how didn't it fail the stress on that region.

https://argus.scylladb.com/test/50e0c80c-ad3b-4c9e-93c7-69b30dd821f2/runs?additionalRuns[]=6ccc34ab-465c-4837-b6a9-ad513d2e8436

Nemesis `NodetoolDrain` was run on the `us-east-1` DC (with 2 nodes), no c-s failures/errors:
```
2023-05-09 00:34:45.233: (DisruptionEvent Severity.NORMAL) period_type=begin event_id=11088950-927d-447b-9ae0-0f407ecf25d7: nemesis_name=NodetoolDrain target_node=Node lwt-longevity-multi-dc-24h-lwt-mult-db-node
-6ccc34ab-8 [18.205.7.166 | 10.12.3.87] (seed: False)
2023-05-09 00:34:46.391: (NodetoolEvent Severity.NORMAL) period_type=begin event_id=e0aaa682-677f-40ca-969d-57ce0bb3b0e6: nodetool_command=drain node=lwt-longevity-multi-dc-24h-lwt-mult-db-node-6ccc34ab-8
2023-05-09T00:34:48+00:00 lwt-longevity-multi-dc-24h-lwt-mult-db-node-6ccc34ab-8     !INFO | scylla[6177]:  [shard  0] rpc - client 10.15.0.101:52560 msg_id 2:  exception "gate closed" in no_wait handler ignored
2023-05-09T00:34:49+00:00 lwt-longevity-multi-dc-24h-lwt-mult-db-node-6ccc34ab-8     !INFO | scylla[6177]:  [shard  0] rpc - client 10.4.2.114:62790 msg_id 2:  exception "gate closed" in no_wait handler ignored
2023-05-09T00:34:49+00:00 lwt-longevity-multi-dc-24h-lwt-mult-db-node-6ccc34ab-8     !INFO | scylla[6177]:  [shard  0] rpc - client 10.4.2.173:62100 msg_id 2:  exception "gate closed" in no_wait handler ignored
2023-05-09T00:34:49+00:00 lwt-longevity-multi-dc-24h-lwt-mult-db-node-6ccc34ab-8     !INFO | scylla[6177]:  [shard  0] rpc - client 10.4.2.114:62790 msg_id 3:  exception "gate closed" in no_wait handler ignored
2023-05-09T00:34:50+00:00 lwt-longevity-multi-dc-24h-lwt-mult-db-node-6ccc34ab-8     !INFO | scylla[6177]:  [shard  0] rpc - client 10.4.0.244:58350 msg_id 2:  exception "gate closed" in no_wait handler ignored
2023-05-09T00:34:50+00:00 lwt-longevity-multi-dc-24h-lwt-mult-db-node-6ccc34ab-8     !INFO | scylla[6177]:  [shard  0] rpc - client 10.15.0.101:52560 msg_id 3:  exception "gate closed" in no_wait handler ignored
2023-05-09T00:34:50+00:00 lwt-longevity-multi-dc-24h-lwt-mult-db-node-6ccc34ab-8     !INFO | scylla[6177]:  [shard  0] rpc - client 10.12.1.49:58590 msg_id 2:  exception "gate closed" in no_wait handler ignored
2023-05-09T00:34:50+00:00 lwt-longevity-multi-dc-24h-lwt-mult-db-node-6ccc34ab-8     !INFO | scylla[6177]:  [shard  0] rpc - client 10.4.2.114:62790 msg_id 4:  exception "gate closed" in no_wait handler ignored
2023-05-09T00:34:51+00:00 lwt-longevity-multi-dc-24h-lwt-mult-db-node-6ccc34ab-8     !INFO | scylla[6177]:  [shard  0] rpc - client 10.4.2.173:62100 msg_id 3:  exception "gate closed" in no_wait handler ignored
2023-05-09T00:34:51+00:00 lwt-longevity-multi-dc-24h-lwt-mult-db-node-6ccc34ab-8     !INFO | scylla[6177]:  [shard  0] rpc - client 10.4.2.114:62790 msg_id 5:  exception "gate closed" in no_wait handler ignored
2023-05-09T00:34:51+00:00 lwt-longevity-multi-dc-24h-lwt-mult-db-node-6ccc34ab-8     !INFO | scylla[6177]:  [shard  0] rpc - client 10.15.1.41:60180 msg_id 2:  exception "gate closed" in no_wait handler ignored
2023-05-09 00:34:53.539: (NodetoolEvent Severity.NORMAL) period_type=end event_id=e0aaa682-677f-40ca-969d-57ce0bb3b0e6 duration=7s: nodetool_command=drain node=lwt-longevity-multi-dc-24h-lwt-mult-db-node-6ccc34a
b-8
2023-05-09 00:34:53.540: (NodetoolEvent Severity.NORMAL) period_type=begin event_id=d96e316b-b1c9-49de-acc0-94937177f23a: nodetool_command=status node=lwt-longevity-multi-dc-24h-lwt-mult-db-node-6ccc34ab-8
2023-05-09 00:34:55.611: (NodetoolEvent Severity.NORMAL) period_type=end event_id=d96e316b-b1c9-49de-acc0-94937177f23a duration=2s: nodetool_command=status node=lwt-longevity-multi-dc-24h-lwt-mult-db-node-6ccc34
ab-8
2023-05-09 00:34:56.323: (JMXServiceEvent Severity.NORMAL) period_type=end event_id=75f07ed8-d8f5-4de9-9352-e15f17b0cf42 duration=11h36m53s node=lwt-longevity-multi-dc-24h-lwt-mult-db-node-6ccc34ab-8
2023-05-09 00:34:56.526: (ScyllaServerStatusEvent Severity.NORMAL) period_type=end event_id=95f1878f-8173-42a0-872b-e91651484ec4 duration=11h37m10s node=lwt-longevity-multi-dc-24h-lwt-mult-db-node-6ccc34ab-8
2023-05-09 00:36:25.119: (ScyllaServerStatusEvent Severity.NORMAL) period_type=begin event_id=2c3abe8c-4e6b-4383-b631-9b485fce3137 node=lwt-longevity-multi-dc-24h-lwt-mult-db-node-6ccc34ab-8
2023-05-09 00:36:33.424: (JMXServiceEvent Severity.NORMAL) period_type=begin event_id=63b037bf-a1ef-4090-b7cc-3d636d7ecb57 node=lwt-longevity-multi-dc-24h-lwt-mult-db-node-6ccc34ab-8

2023-05-09 00:37:35.331: (DisruptionEvent Severity.NORMAL) period_type=end event_id=11088950-927d-447b-9ae0-0f407ecf25d7 duration=2m50s: nemesis_name=NodetoolDrain target_node=Node lwt-longevity-multi-dc-24h-lwt-mult-db-node-6ccc34ab-8 [18.205.7.166 | 10.12.3.87] (seed: False)
```

`ModifyTable` nemesis - no c-s failures/errors:
```
2023-05-09 00:56:27.783: (DisruptionEvent Severity.NORMAL) period_type=begin event_id=67ec3606-9551-4500-ad65-f28a5e3439e1: nemesis_name=ModifyTable target_node=Node lwt-longevity-multi-dc-24h-lwt-mult-db-node-6
ccc34ab-9 [18.234.200.47 | 10.12.1.49] (seed: False)
2023-05-09 00:56:28.902: (InfoEvent Severity.NORMAL) period_type=not-set event_id=4d71463d-561e-476d-9bde-e59dc9838ebc: message=ModifyTablePropertiesComment Node lwt-longevity-multi-dc-24h-lwt-mult-db-node-6ccc3
4ab-9 [18.234.200.47 | 10.12.1.49] (seed: False)
2023-05-09 00:57:23.609: (DisruptionEvent Severity.NORMAL) period_type=end event_id=67ec3606-9551-4500-ad65-f28a5e3439e1 duration=55s: nemesis_name=ModifyTable target_node=Node lwt-longevity-multi-dc-24h-lwt-mul
t-db-node-6ccc34ab-9 [18.234.200.47 | 10.12.1.49] (seed: False)
```

Also `disrupt_destroy_data_then_repair` was run on the node from this DC

Reply:
How do we know that there was a loader connected exactly to that region?

Reply:
> How do we know that there was a loader connected exactly to that region?

We create one loader in region. It's defined in the test yaml:
`n_loaders: '1 1 1'

and also add `datacenter` parameter with DC name of the loader to the cassandra-stress command:
```
cassandra-stress user no-warmup profile=/tmp/c-s_lwt_big_data_multidc.yaml n=10000000 ops'"'"'(insert_query=1)'"'"' cl=LOCAL_QUORUM 
serial-cl=LOCAL_SERIAL -mode native cql3 -rate threads=1000 -transport '"'"'truststore=/etc/scylla/ssl_conf/client/cacerts.jks truststore-password=cassandra'"'"' -node datacenter=us-eastscylla_node_east 10.4.2.1
14,10.4.0.12,10.4.2.173,10.4.0.244,10.15.0.45,10.15.0.101,10.15.3.72,10.12.3.87,10.12.1.49 -errors skip-unsupported-columns'
```

when loader is in us-east-1:
```
(CassandraStressEvent Severity.NORMAL) period_type=begin event_id=f9587627-d2a5-445f-a16d-1ad3f84a0e7f: node=Node lwt-longevity-multi-dc-24h-lwt-mult-loader-node-6ccc34ab-3 [44.201.14.135 | 10.12.1.112]
```

Reply:
So, we need to understand it better.
Something doesn't do what we think it should do.

Reply:
datacenter=us-eastscylla_node_east

Should force the c-s to use only that datacenter, regardless of the list of contact points we pass to it.

Reply:
> datacenter=us-eastscylla_node_east
> 
> Should force the c-s to use only that datacenter, regardless of the list of contact points we pass to it.

Need to take another look at drivers docs, if it's just to prefer this datacenter,  then we don't have a problem.

Reply:
> > > How do we know that there was a loader connected exactly to that region?
> > 
> > 
> > We create one loader in region. It's defined in the test yaml: `n_loaders: '1 1 1'
> > and also add `datacenter` parameter with DC name of the loader to the cassandra-stress command:
> > ```
> > cassandra-stress user no-warmup profile=/tmp/c-s_lwt_big_data_multidc.yaml n=10000000 ops'"'"'(insert_query=1)'"'"' cl=LOCAL_QUORUM 
> > serial-cl=LOCAL_SERIAL -mode native cql3 -rate threads=1000 -transport '"'"'truststore=/etc/scylla/ssl_conf/client/cacerts.jks truststore-password=cassandra'"'"' -node datacenter=us-eastscylla_node_east 10.4.2.1
> > 14,10.4.0.12,10.4.2.173,10.4.0.244,10.15.0.45,10.15.0.101,10.15.3.72,10.12.3.87,10.12.1.49 -errors skip-unsupported-columns'
> > ```
> 
> I see 9 IP addresses here which makes me think it connects to all the DB cluster nodes.

This is nodes list that we transfer in the command. But because the default load balancing policy in scylla-tools-java is DCAwareRoundRobinPolicy which prefers nodes in the local datacenter, the requests are sent to one DC only, where loader is located.

From cassandra-stress log:

WARN  14:07:34,332 **Some contact points don't match local data center. Local DC = us-eastscylla_node_east. Non-conforming contact points**: /10.4.2.114:9042 (eu-westscylla_node_west),/10.15.0.45:9042 (us-west-2scylla_node_west),/10.4.0.244:9042 (eu-westscylla_node_west),/10.15.3.72:9042 (us-west-2scylla_node_west),/10.15.0.101:9042 (us-west-2scylla_node_west),/10.4.0.12:9042 (eu-westscylla_node_west),/10.4.2.173:9042 (eu-westscylla_node_west)

> 
> > when loader is in us-east-1:
> > ```
> > (CassandraStressEvent Severity.NORMAL) period_type=begin event_id=f9587627-d2a5-445f-a16d-1ad3f84a0e7f: node=Node lwt-longevity-multi-dc-24h-lwt-mult-loader-node-6ccc34ab-3 [44.201.14.135 | 10.12.1.112]
> > ```
> 
> The placement of the loader doesn't force it to connect to the nodes of only the same region.

It is exacly what happens because of DCAwareRoundRobinPolicy

Reply:
> > datacenter=us-eastscylla_node_east
> > Should force the c-s to use only that datacenter, regardless of the list of contact points we pass to it.
> 
> Need to take another look at drivers docs, if it's just to prefer this datacenter, then we don't have a problem.

I see it in the cassandra-stress log - the local DC is prefered:

WARN 14:07:34,332 **Some contact points don't match local data center. Local DC = us-eastscylla_node_east. Non-conforming contact points**: /10.4.2.114:9042 (eu-westscylla_node_west),/10.15.0.45:9042 (us-west-2scylla_node_west),/10.4.0.244:9042 (eu-westscylla_node_west),/10.15.3.72:9042 (us-west-2scylla_node_west),/10.15.0.101:9042 (us-west-2scylla_node_west),/10.4.0.12:9042 (eu-westscylla_node_west),/10.4.2.173:9042 (eu-westscylla_node_west)

Reply:
It can be checked by blocking the IP addresses of the DB nodes from the foreign regions on each of the loaders.